[flang][debug] Support derived type components with box types.

abidh opened the pull request:
Our support for derived types uses `getTypeSizeAndAlignment` to calculate the offset of the members. The `fir.box` was not supported in that function. It meant  that any member which required descriptor was not supported in the derived type.
    
This PR enhances `getTypeSizeAndAlignment` to support `fir.box`. The implementation uses the following formula to calculate the size.
    size = offset_of_dims_field + (rank * size_of_dims_field)
If an addendum is required then size of its 2 field is also added to the final size.
    
The code to calculate the offset of the descriptor was already present in the `DebugTypeGenerator.cpp`. It has been moved to a separate file so that it can be used in multiple places.
    
There are 2 other changes in this PR:
1. The `recID` field is used to handle cases where we have a member references its parent type.
2. A type cache is maintained to avoid duplication. It is also needed for circular reference case.

I tested it with the following derived type and the values of all the fields were correct in the debugger. The `debug-cyclic-derived-type.f90` which tests the cyclic dependency has been updated and works as expected. 

```
  type :: sometype
    integer :: m_array(3) = 42
    type(t1), allocatable :: m_vt1
    integer v2
    integer, allocatable :: m_alloc(:)
    integer v3
    character(len=:), allocatable :: m_first
    integer v4
    integer, pointer :: m_p1 => my_target
    integer v5
    integer, pointer :: m_p2! => NULL()
    integer v6
    integer, pointer :: m_p3(:)
    integer v7
  end type
```


Fixes #108001.

abidh committed:
[flang][debug] Support derived type components with box types.

Our support for derived types uses `getTypeSizeAndAlignment` to calculate
the offset of the members. The `fir.box` was not supported in that
function. It meant  that any member which required descriptor was not
supported in the derived type.

This PR enhances `getTypeSizeAndAlignment` to support `fir.box`. The
implementation uses the following formula to calculate the size.
size = offset_of_dims_field + (rank * size_of_dims_field)
If an addendum is required then size of its 2 field is also added to the
final size.

The code to calculate the offset of the descriptor was already present
in the `DebugTypeGenerator.cpp`. It has been moved to a separate file
so that it can be used in multiple places.

There are 2 other changes in this PR:
1. The recID field is used to handle cases where we have a member
references its parent type.
2. A type cache is maintained to avoid duplication. It is also needed
for circular reference case.

Fixes #108001.

abidh committed:
Update cyclic derived type test.

pawosm-arm commented on flang/test/Integration/debug-cyclic-derived-type.f90:
```diff
@@ -11,5 +11,11 @@ module m
+
+! CHECK-DAG: ![[T2]] = {{.*}}!DICompositeType(tag: DW_TAG_structure_type, name: "t2"{{.*}}elements: ![[T2_ELEMS:[0-9]+]])
+! CHECK-DAG: ![[T2_ELEMS]] = !{![[T2_ELEM1:[0-9]+]]}
+! CHECK-DAG: ![[T2_ELEM1]] = !DIDerivedType(tag: DW_TAG_member, name: "v1", baseType: ![[T1]]{{.*}})
```
NIT: `\ No newline at end of file` reported by git for this file

abidh replied:
Thanks for catching this. Fixed now.

abidh committed:
Add new line at the end of file.

pawosm-arm reviewed (approved):
I've tested it locally, it works, thanks for fixing this

pawosm-arm commented on flang/include/flang/Optimizer/CodeGen/DescriptorOffsets.h:
```diff
@@ -0,0 +1,39 @@
+static std::uint64_t getDescComponentOffset(const mlir::DataLayout &dl,
+                                            mlir::MLIRContext *context,
+                                            mlir::Type fieldType) {
+  static_assert(Field > 0 && Field < 8);
```
NB, Is there any constant defined that could be used here instead of 8?

jeanPerier commented on flang/lib/Optimizer/Dialect/FIRType.cpp:
```diff
@@ -13,6 +13,9 @@
 #include "flang/Optimizer/Builder/Todo.h"
+#include "flang/Optimizer/CodeGen/DescriptorModel.h"
+#include "flang/Optimizer/CodeGen/DescriptorOffsets.h"
+#include "flang/Optimizer/CodeGen/TypeConverter.h"
```
So far I was rather against handling fir.box<> into `getTypeSizeAndAlignment` because to me it is meant to be an abstract type as much possible (could be implemented as a CFI descriptor, or some descriptor compatible with existing compilers if we wanted). Instead, I was more in favor of translating type to LLVM and using `getTypeSizeAndAlignment` at that level. That way, the codegen of abstract FIR types is centralized in the type codegen and needs not to be spilled in other places.

Now I understand this is inconvenient given there are several places that needs to know about the size at different stage of the compilation.

But as far as debugging is concerned, I would still be in favor of calling translating types to LLVM and calling `getTypeSizeAndAlignment` with that.

Otherwise, we at least need to centralize all low level descriptor related tools in some flang/Optimizer/Dialect/Support/DescriptorModel.h file. The Dialect library should not depend upon Codegen.

abidh replied:
Thanks for your comments. I have updated the PR and it does not use `getTypeSizeAndAlignment` to get the size for any type. We use LLVMTypeConverter to convert fieldTy into LLVM type and then use DataLayout to get the size. It all seems to works ok. My concern was whether it is safe to call `DataLayout::getTypeSize` on any type that is returned by `LLVMTypeConverter::convertType`. I did not see any problem with my testing but would be happy to add any type specific check.

jeanPerier commented on flang/lib/Optimizer/Dialect/FIRType.cpp:
```diff
@@ -1455,6 +1458,31 @@ fir::getTypeSizeAndAlignment(mlir::Location loc, mlir::Type ty,
+    mlir::Type dimsType = getDescFieldTypeModel<kDimsPosInBox>()(context);
+    std::uint64_t size =
+        getDescComponentOffset<kDimsPosInBox>(dl, context, dimsType);
+    unsigned rank = getBoxRank(ty);
```
What about assumed-ranks, the descriptor size is not a compile time constant? They cannot be used as derived type components, but this generic helper would need to handle them or bail correctly.

jeanPerier reviewed (commented):
fir.box type is rea

abidh committed:
Handle review comments.

Use LLVMTypeConverter to convert the fir.box into an mlir::Type and then get its size from DataLayout.

jeanPerier commented on flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:
```diff
@@ -163,26 +163,45 @@ mlir::LLVM::DITypeAttr DebugTypeGenerator::convertBoxedSequenceType(
 
-  llvm::SmallVector<mlir::LLVM::DINodeAttr> elements;
   std::uint64_t offset = 0;
+  LLVMTypeConverter llvmTypeConverter(module, false, false, *dataLayout);
```
This is likely not a trivial class/ctor, and I would not be surprised if it did caching somewhere. Probably best to keep it at the same level as the dataLayout.

jeanPerier reviewed (approved):
Thanks for the simplification, LGTM. Getting the size on the LLVM type is safe (the only trick/ugliness is to take care of the fir.box translation to LLVM that is not context free, and you did take care of that).

abidh committed:
Make llvmTypeConverter a class member.

This helps avoid its instantiation for every RecordType.

abidh committed:
Workaround limitation in mlir handling of cyclic debug attributes.

This commit fixes an issue that can cause an assertion to fail in some
circumstances at
https://github.com/llvm/llvm-project/blob/26029d77a57cb4aaa1479064109e985a90d0edd8/mlir/lib/Target/LLVMIR/DebugTranslation.cpp#L270.
The issue relates to the handling of recursive debug type in mlir.
See the discussion at the end of follow PR for more details.
https://github.com/llvm/llvm-project/pull/106571

Problem could be explained with the following example code:

type t2
  type(t1), pointer :: p1
end type

type t1
  type(t2), pointer :: p2
end type

In the description below, type_self means a temporary type that is
generated as a place holder while the members of that type are being
processed.

If we process t1 first then we will have the following structure after
it has been processed.
 t1 -> t2 -> t1_self

This is because when we started processing t2, we did not have the
complete t1 but its place holder t1_self. Now if some entity requires
t2, we will already have that in cache and will return it. But this t2
refers to t1_self and not to t1.

In mlir handling, only those types are allowed to have _self reference
which are wrapped by entity whose reference it contains. So
t1 -> t2 -> t1_self is ok because the t1_self reference can be resolved
by the outer t1. But standalone t2 is not because there will be no way
to resolve it. Please see DebugTranslation::translateRecursive for
details on how mlir handles recursive types.

The fix is not to cache the type that will fail if used standalone.

abidh commented:
While testing, I noticed an assertion failure [at the following location](https://github.com/llvm/llvm-project/blob/26029d77a57cb4aaa1479064109e985a90d0edd8/mlir/lib/Target/LLVMIR/DebugTranslation.cpp#L270). I got the following [advice](https://github.com/llvm/llvm-project/pull/106571#issuecomment-2371931898) when I asked how best to handle the issue with the author of the code. I have implemented this in this [commit](https://github.com/llvm/llvm-project/pull/109424/commits/a45bef35ea17e91a62adedffc6604d0861bd14fd) by not caching a type if we know that its standalone use could cause a problem. This will result a new type attribute to be generated for that type if it is used standalone.
